### PR TITLE
Added KPSS test (1)

### PR DESCRIPTION
**Pull request by N-Wouda**
Full implementation of the KPSS test; with unit tests and description in the
docs. See #2749 for a lengthy discussion on the design choices.

### CODE REVIEW

**Comment by N-Wouda** [comment]
@josef-pkt here's the 'new' PR, with everything bundled into a single commit.

File: statsmodels/tsa/stattools.py [hunk 1]
```diff
@@ -1066,12 +1194,12 @@ def arma_order_select_ic(y, max_ar=4, max_ma=2, ic='bic', trend='c',
...
-    acf3, ci3, Q3, pvalue3 = acf(x, nlags=40, confint=95, qstat=True,
-                                 unbiased=True)
-    acf4, ci4, Q4, pvalue4 = acf(x, nlags=40, confint=95, fft=True, qstat=True,
-                                 unbiased=True)
+#    acf1, ci1, Q, pvalue = acf(x, nlags=40, confint=95, qstat=True)
+#    acf2, ci2, Q2, pvalue2 = acf(x, nlags=40, confint=95, fft=True, qstat=True)
+#    acf3, ci3, Q3, pvalue3 = acf(x, nlags=40, confint=95, qstat=True,
+#                                 unbiased=True)
+#    acf4, ci4, Q4, pvalue4 = acf(x, nlags=40, confint=95, fft=True, qstat=True,
+#                                 unbiased=True)
```
Comment (N-Wouda) [inline statsmodels/tsa/stattools.py]: Commented out, since `acf` no longer takes a `confint` argument. This was in a MAINT note.

**Comment by josef-pkt** [comment]
The branch looks good.

3 more comments with small issues (not strictly necessary)

- warnings.warn will warn only the first time it is used. In general we use custom warning that are set to always. We don't have a warning class yet, that would match this https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tools/sm_exceptions.py

- no plain `assert` in unit tests, numpy testing has an `assert_` function for this. (It's more informative and doesn't get removed if someone compiles the python code with `-o` option.)
- unittest of resultsstore: it would be better to test for correct attributes instead of `isinstance`. That the class is ResultsStore is not relevant and not guaranteed in future refactorings.

**Comment by N-Wouda** [comment]
@josef-pkt,

* Would `OutOfBoundsInterpolationWarning` make sense? I think it's appropriate. If you agree, I'll add it in and use that instead of a generic `UserWarning`.

* I'll switch to `assert_`.

* I'll test for some attributes instead.

**Comment by josef-pkt** [comment]
`OutOfBoundsInterpolationWarning`  is too specific and there wouldn't be many use cases for it.  
I thought we could use something more generic but I didn't come up with anything (yet). In the one or few other cases like this, we only write in the docstring that the pvalue is at the boundary point of the available table, but don't warn of flag.

IIRC, Kevin did some large scale simulations to create pvalues or critical values for unit root tests, so eventually we might be able to replace simple tables of critical values with fancier methods.

**Comment by N-Wouda** [comment]
@josef-pkt we can always leave the issue of a suitable warning as something for the future. I could mention it alongside suitable lag selection in the issue I will open for the KPSS test after merger (as discussed previously in #2749). 

I'll push a single more commit to this PR to deal with the `assert` and testing for attributes in the unit test instead (that will probably be tomorrow), and then I think we may be good to go!

**Comment by josef-pkt** [comment]
@N-Wouda   
BTW: Are you aware that we have an almost finished (but stalled) PR for exponential smoothing? #1489

**Comment by N-Wouda** [comment]
@josef-pkt had not seen that before, no! I did see some feature requests for it, but did not realise someone was already on it. Is it based on `statespace`?

With regard to this PR, I think I pretty much got it now, besides the warning message. I'll leave that in an issue to be addressed in future PR's (together with lags selection).

Starting February, I'll probably get onto either bootstrapped CI, or exponential smoothing. I have a lot of books laying around on the latter over here (including a couple by Hyndman), so that may be more convenient. Do you have any preferences on this, as in, what would you like to see first? :).

**Comment by josef-pkt** [comment]
About ETS:  I haven't looked much at the details. But the PR and code is older than the statespace models that Chad added over the last 1 to 2 years.
From reading or skimming some of Hyndman's writing, I think most versions have a linear statespace or ARMA representation, but I'm not sure all do. I don't know what the advantage would be of the current standalone version and a version written on top of the new tsa.statespace. 
(My guess is that the special implementation would use less memory because it doesn't need to get the generic representation for long AR/seasonal components, and potentially faster if it doesn't use generic calculations that are redundant in this case. just guessing)

> Do you have any preferences on this, as in, what would you like to see first?
Both are fine with me.
Based on user demand I would pitch for ETS. Based on the description in PR #1489, it is almost finished, needs some decisions on or options for initial conditions, and maybe examples and tests.
With not too much effort we would get a first version that covers many specifications.
(I don't know whether it would also be quick and easy to implement it on top of tsa.statespace. The framework that Chad built is pretty extensive already, so many things could be reused for new linear models.)

**Comment by N-Wouda** [comment]
@josef-pkt I am more than willing to take a look at the ETS in #1489 sometime soon. I will probably bug Chad with it, since I may have some questions about `tsa.statespace`.

As to this PR, is it ready for a merge? I will move the ETS discussion to the appropriate PR.

**Comment by josef-pkt** [comment]
LGTM, merging

@N-Wouda  Thank you

**Comment by josef-pkt** [comment]
@ChadFulton   Thanks for reviewing, that made this a lot faster

**Comment by ChadFulton** [comment]
No problem, I'm trying to get more involved when possible.

**Pull request merged**